File: api/flpc/store.py
"""SQLite store: (re)build from the FLOIR workbooks and cached dimensions.

The server watches its input folders; when a workbook (or the carrier-group
CSV) is added or changed it rebuilds the store with `etl/ingest.py` and swaps
it in atomically. Drop a new quarter's .xlsx in the folder — no restart.
"""

from __future__ import annotations

import glob
import importlib.util
import json
import logging
import os
import sqlite3
import threading
import time
from dataclasses import dataclass

from . import config

log = logging.getLogger("flpc.store")


def _load_ingest():
    path = os.path.join(config.REPO_ROOT, "etl", "ingest.py")
    spec = importlib.util.spec_from_file_location("flpc_ingest", path)
    mod = importlib.util.module_from_spec(spec)
    spec.loader.exec_module(mod)
    return mod


@dataclass
class Company:
    naic: str
    name: str
    names: list[str]
    group_id: str
    group_name: str
    first: str
    last: str


@dataclass
class PolicyType:
    id: str
    name: str
    line: str
    product: str
    wind_only: bool


class Store:
    def __init__(self, input_dirs=None, db_path=None, groups_path=None, reload_interval=None):
        self.input_dirs = input_dirs or config.INPUT_DIRS
        self.db_path = db_path or config.DB_PATH
        self.groups_path = groups_path or config.groups_path()
        self.reload_interval = config.RELOAD_INTERVAL if reload_interval is None else reload_interval
        self._lock = threading.Lock()
        self._last_check = 0.0
        self._sig = None
        self.version = 0
        self.last_error: str | None = None
        self.ensure(force=True)

    # ------------------------------------------------------------------ #
    # build / reload
    # ------------------------------------------------------------------ #
    def _signature(self):
        items = []
        for d in self.input_dirs:
            for p in sorted(glob.glob(os.path.join(d, "*.xlsx"))):
                st = os.stat(p)
                items.append((os.path.basename(p), int(st.st_mtime), st.st_size))
        for p in (self.groups_path, os.path.join(config.REPO_ROOT, "etl", "ingest.py")):
            if os.path.exists(p):
                st = os.stat(p)
                items.append((p, int(st.st_mtime), st.st_size))
        return json.dumps(items)

    def ensure(self, force: bool = False):
        """Rebuild the store if inputs changed (checked at most every N s)."""
        now = time.time()
        if not force and now - self._last_check < self.reload_interval:
            return
        with self._lock:
            if not force and now - self._last_check < self.reload_interval:
                return
            self._last_check = now
            sig = self._signature()
            sig_file = self.db_path + ".sig"
            if sig == self._sig and os.path.exists(self.db_path):
                return
            stored = open(sig_file).read() if os.path.exists(sig_file) else None
            has_xlsx = '.xlsx"' in sig
            if os.path.exists(self.db_path) and (stored == sig or not has_xlsx):
                pass                                   # up to date (or nothing to build from)
            else:
                try:
                    log.info("building store from %s", self.input_dirs)
                    ingest = _load_ingest()
                    data = ingest.build(self.input_dirs, self.groups_path)
                    ingest.write_sqlite(data, self.db_path)
                    with open(sig_file, "w") as fh:
                        fh.write(sig)
                    self.last_error = None
                    log.info("store built: %d periods, %d rows", len(data["period_order"]), len(data["facts_b"]))
                except BaseException as e:            # keep serving the previous store
                    self.last_error = f"rebuild failed: {e}"
                    log.exception("store rebuild failed")
                    if not os.path.exists(self.db_path):
                        raise RuntimeError(self.last_error) from e
            self._sig = sig
            self._load_dims()

    def connect(self) -> sqlite3.Connection:
        con = sqlite3.connect(f"file:{self.db_path}?mode=ro", uri=True, check_same_thread=False)
        con.row_factory = sqlite3.Row
        return con

    # ------------------------------------------------------------------ #
    # cached dimensions
    # ------------------------------------------------------------------ #
    def _load_dims(self):
        con = self.connect()
        try:
            q = lambda sql: con.execute(sql).fetchall()
            self.meta = {r["key"]: r["value"] for r in q("SELECT * FROM meta")}
            self.periods = [(r["period"], r["idx"]) for r in q("SELECT period, idx FROM periods ORDER BY idx")]
            self.period_idx = dict(self.periods)
            self.idx_period = {i: p for p, i in self.periods}
            self.metric_cols = [r[1] for r in q("PRAGMA table_info(facts)")][8:]
            self.metric_avail = {r["metric"]: (self.period_idx[r["first_period"]], self.period_idx[r["last_period"]])
                                 for r in q("SELECT * FROM metrics")}
            self.companies = {
                r["naic"]: Company(r["naic"], r["name"], r["names"].split(" | ") if r["names"] else [],
                                   r["group_id"], r["group_name"], r["first_period"], r["last_period"])
                for r in q("SELECT * FROM companies")}
            self.groups = {r["group_id"]: {"name": r["group_name"], "standalone": bool(r["standalone"]),
                                           "members": r["members"].split(",")}
                           for r in q("SELECT * FROM groups")}
            self.policy_types = {r["pt_id"]: PolicyType(r["pt_id"], r["policy_type"], r["line"], r["product"],
                                                        bool(r["wind_only"]))
                                 for r in q("SELECT * FROM policy_types ORDER BY pt_id")}
        finally:
            con.close()
        self.version += 1

    @property
    def latest_idx(self) -> int:
        return self.periods[-1][1]

    def available(self, metric: str, idx: int) -> bool:
        a = self.metric_avail.get(metric)
        return bool(a) and a[0] <= idx <= a[1]

File: supabase/functions/_shared/flpc/store.ts
// Postgres-backed store (hosted counterpart of api/flpc/store.py).
//
// The data tables live in the `flpc` schema and are replaced as a whole by
// flpc.load_bundle(). Small dimensions (periods, companies, groups, policy
// types, metric availability) are cached per function instance and reloaded
// when meta.load_id changes, which a cheap one-row read detects.

export interface Db {
  /** Rows as arrays, in select-list order. */
  values(sql: string, params?: unknown[]): Promise<unknown[][]>;
  /** Rows as objects. */
  rows<T = Record<string, unknown>>(sql: string, params?: unknown[]): Promise<T[]>;
}

export interface Company {
  naic: string;
  name: string;
  names: string[];
  group_id: string;
  group_name: string;
  first: string;
  last: string;
}

export interface Group {
  name: string;
  standalone: boolean;
  members: string[];
}

export interface PolicyType {
  id: string;
  name: string;
  line: string;
  product: string;
  wind_only: boolean;
}

/** Non-metric columns of flpc.facts; every other column is a metric. */
export const FACT_DIMS = ["period", "idx", "naic", "group_id", "pt_id", "line", "product", "wind_only"];

const RECHECK_MS = 5_000;

export class Store {
  db: Db;
  loadId: string | null = null;
  version = 0;
  lastError: string | null = null;
  meta: Record<string, string> = {};
  periods: [string, number][] = [];
  periodIdx = new Map<string, number>();
  idxPeriod = new Map<number, string>();
  metricCols: string[] = [];
  metricAvail = new Map<string, [number, number]>();
  companies = new Map<string, Company>();
  groups = new Map<string, Group>();
  policyTypes = new Map<string, PolicyType>();
  sizeCache: [number, Map<string, number>] | null = null;
  private checkedAt = 0;
  private loading: Promise<void> | null = null;

  constructor(db: Db) {
    this.db = db;
  }

  get latestIdx(): number {
    return this.periods[this.periods.length - 1][1];
  }

  get loaded(): boolean {
    return this.periods.length > 0;
  }

  available(metric: string, idx: number): boolean {
    const a = this.metricAvail.get(metric);
    return !!a && a[0] <= idx && idx <= a[1];
  }

  /** Make sure the cached dimensions match the loaded data (checked every few seconds). */
  async ensure(force = false): Promise<this> {
    const now = Date.now();
    if (!force && this.loaded && now - this.checkedAt < RECHECK_MS) return this;
    if (this.loading) {
      await this.loading;
      return this;
    }
    this.loading = (async () => {
      const r = await this.db.rows<{ value: string }>("SELECT value FROM flpc.meta WHERE key = 'load_id'");
      const lid = r[0]?.value ?? null;
      if (force || lid !== this.loadId || !this.loaded) await this.loadDims(lid);
      this.checkedAt = Date.now();
    })();
    try {
      await this.loading;
    } finally {
      this.loading = null;
    }
    return this;
  }

  /** Forget the cache (after a data load in this instance). */
  invalidate() {
    this.checkedAt = 0;
    this.loadId = "\0stale";
  }

  private async loadDims(lid: string | null) {
    const db = this.db;
    const [meta, periods, cols, avail, comps, groups, pts] = await Promise.all([
      db.rows<{ key: string; value: string }>("SELECT key, value FROM flpc.meta"),
      db.rows<{ period: string; idx: number }>("SELECT period, idx FROM flpc.periods ORDER BY idx"),
      db.rows<{ column_name: string }>(
        "SELECT column_name FROM information_schema.columns WHERE table_schema = 'flpc' AND table_name = 'facts' " +
          "ORDER BY ordinal_position"),
      db.rows<{ metric: string; first_period: string; last_period: string }>(
        "SELECT metric, first_period, last_period FROM flpc.metrics"),
      db.rows<Record<string, string>>(
        "SELECT naic, name, names, group_id, group_name, first_period, last_period FROM flpc.companies ORDER BY ord"),
      db.rows<{ group_id: string; group_name: string; standalone: number; members: string }>(
        "SELECT group_id, group_name, standalone, members FROM flpc.groups ORDER BY ord"),
      db.rows<{ pt_id: string; policy_type: string; line: string; product: string; wind_only: number }>(
        `SELECT pt_id, policy_type, line, product, wind_only FROM flpc.policy_types ORDER BY pt_id COLLATE "C"`),
    ]);
    if (!periods.length) throw new Error("no data loaded yet: run the ETL with --push (see docs/SUPABASE.md)");
    this.meta = Object.fromEntries(meta.map((r) => [r.key, r.value]));
    this.periods = periods.map((r) => [r.period, r.idx]);
    this.periodIdx = new Map(this.periods);
    this.idxPeriod = new Map(this.periods.map(([p, i]) => [i, p]));
    this.metricCols = cols.map((r) => r.column_name).filter((c) => !FACT_DIMS.includes(c));
    this.metricAvail = new Map(avail.map((r) => [r.metric, [this.periodIdx.get(r.first_period)!,
      this.periodIdx.get(r.last_period)!]]));
    this.companies = new Map(comps.map((r) => [r.naic, {
      naic: r.naic, name: r.name, names: r.names ? r.names.split(" | ") : [], group_id: r.group_id,
      group_name: r.group_name, first: r.first_period, last: r.last_period,
    }]));
    this.groups = new Map(groups.map((r) => [r.group_id, {
      name: r.group_name, standalone: !!r.standalone, members: r.members.split(","),
    }]));
    this.policyTypes = new Map(pts.map((r) => [r.pt_id, {
      id: r.pt_id, name: r.policy_type, line: r.line, product: r.product, wind_only: !!r.wind_only,
    }]));
    this.loadId = lid;
    this.sizeCache = null;
    this.version++;
  }
}
